kobo/apps/openrosa/apps/api/tools.py

```diff
@@ -157,22 +160,22 @@ def get_media_file_response(
         return HttpResponseRedirect(metadata.data_value)
 
     # When `request.user` is authenticated, their authentication is lost with
-    # an HTTP redirection. We use KoBoCAT to proxy the response from KPI
-    headers = {}
-    if not request.user.is_anonymous:
-        headers = get_request_headers(request.user.username)
-
-    # Send the request internally to avoid extra traffic on the public interface
-    internal_url = metadata.data_value.replace(
-        settings.KOBOFORM_URL, settings.KOBOFORM_INTERNAL_URL
-    )
-    response = requests.get(internal_url, headers=headers)
-
-    return HttpResponse(
-        content=response.content,
-        status=response.status_code,
-        content_type=response.headers['content-type'],
-    )
+    # an HTTP redirection. We need to call KPI viewset directly
+    internal_url = metadata.data_value.replace(settings.KOBOFORM_URL, '')
+    try:
+        resolver_match = resolve(internal_url)
+    except Resolver404:
+        return HttpResponseNotFound()
+
+    args = resolver_match.args
+    kwargs = resolver_match.kwargs
+
+    paired_data_viewset = PairedDataViewset.as_view({'get': 'external'})
+    django_http_request = request._request
+    paired_data_viewset.cls.permission_classes = [XMLExternalDataPermission]
+    paired_data_viewset.cls.renderer_classes = [SubmissionXMLRenderer]
+    paired_data_viewset.cls.filter_backends = []
+    return paired_data_viewset(request=django_http_request, *args, **kwargs)
```

Comment:
I'm a little nervous about having this kind of manipulation of a KPI viewset inside here. It's an unexpected place for access control logic; I fear people reading `kpi/views/v2/paired_data.py` would think they had a complete picture of how the access control works without thinking to look here.

Could you subclass `PairedDataViewset` in `kpi/views/v2/paired_data.py` with the attributes you need and then use that subclass here?